### PR TITLE
filter_irrelevant_rules: Do not filter with every literal/regex

### PR DESCRIPTION
**Pull request by IagoAbal**
Due to constant folding, a rule may be able to match a string that does
not actually exist as-is in the target file, but that it is produced by
propagating constants. E.g., a rule may succeed at matching "foobar" if
the target code contains `"foo" + "bar"`, but a "foobar" filter would
skip the file!

Closes PA-3110

test plan:
make test # added new tests
Plus ran p/default-v2 on 46 repos and on average runtime and memory usage were essentially the same. For five repos, there was a more significant slowdown of about 20%, due to more files being analyzed. There is a large increase in "partially scanned files" because more files are being analyzed, and we encounter some parsing errors on them. This is expected and not bad per se.

### CODE REVIEW

File: cli/src/semgrep/constants.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 DEFAULT_SEMGREP_CONFIG_NAME = "semgrep"
 DEFAULT_SEMGREP_APP_CONFIG_URL = "api/agent/deployments/scans/config"
 
-DEFAULT_TIMEOUT = 30  # seconds
+DEFAULT_TIMEOUT = 2  # seconds
```
Comment (aryx) [inline cli/src/semgrep/constants.py]: we might want to do that in a separate PR.
It's a pretty important change, deserving its own changelog.

File: src/core/Metavariable.ml [hunk 2]
```diff
@@ -34,6 +34,12 @@ let logger = Logging.get_logger [ __MODULE__ ]
  *)
 type mvar = string [@@deriving show, eq, hash]
 
+module MvarSet = Set.Make (struct
```
Comment (aryx) [inline src/core/Metavariable.ml]: You can also use 't Set_.t' (Set_.ml is a polymorphic Set in commons/), or SSet.t (defined in Common.mli I think)

File: cli/src/semgrep/constants.py (hunk 1 above)
Comment (IagoAbal) [reply cli/src/semgrep/constants.py]: Agree, see #8879.

File: src/core/Metavariable.ml (hunk 2 above)
Comment (IagoAbal) [reply src/core/Metavariable.ml]: I am bit fond of defining "new types", so I think it would be nice that `mvar` was not equivalent to `string`, but right now they are plain strings so I'll just go with `StringSet` from `Common2` as suggested.

File: src/optimizing/Analyze_pattern.ml [hunk 3]
```diff
@@ -85,14 +89,35 @@ let extract_strings_and_mvars ?lang any =
...
+              (_, [ Arg { e = L (String (_, (str, _tok), _)); _ } ], _) ) ->
             if not (Pattern.is_special_string_literal str) then
               Common.push str strings
         | IdSpecial (Eval, t) ->
             if Tok.is_origintok t then
               Common.push (Tok.content_of_tok t) strings
-        (* do not recurse there, the type does not have to be in the source *)
-        | TypedMetavar _ -> ()
+        | New (t, _, _, _) when lang = Some Lang.Java ->
+            (* We expect the 'new' keyword to be present.  *)
```
Comment (aryx) [inline src/optimizing/Analyze_pattern.ml]: meh, not super useful I think, most Java programs use 'new'.
In the same way if your pattern contains if/then/else/... we usually don't extract the strings for them because
they are too common.

But maybe it helps, I didn't run any benchmark.

File: src/optimizing/Analyze_pattern.ml [hunk 4]
```diff
@@ -85,14 +89,35 @@ let extract_strings_and_mvars ?lang any =
          * We do now semantic equivance on integers between values so
          * 1000 is now equivalent to 1_000 so we can't "regexpize" it.
          *)
-        | L (String (_, (str, _tok), _)) ->
```
Comment (aryx) [inline src/optimizing/Analyze_pattern.ml:88]: Keep the L (String _) case here, and do nothing, but add a comment saying
we used to extract the string, but we don't because of constant propagation, with the
example you put in the changelog.d

File: src/optimizing/Analyze_rule.mli [hunk 5]
```diff
@@ -37,7 +37,8 @@ val regexp_prefilter_of_rule :
   cache:prefilter_cache option -> Rule.t -> prefilter option
 
 (* internal, do not use directly, not memoized *)
-val regexp_prefilter_of_formula : Rule.formula -> prefilter option
+val regexp_prefilter_of_formula :
```
Comment (aryx) [inline src/optimizing/Analyze_rule.mli:51]: maybe explain why the hell we pass an xlang here. It's not super intuitive.

File: tests/irrelevant_rules/mvar-regex.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+using System;
+
+public class Test
+{
+    public static void Main()
+    {
+        x.SqlFooBar = string.Format("foobar");
```
Comment (aryx) [inline tests/irrelevant_rules/mvar-regex.cs:10]: maybe a comment here explain why it's good we don't match.

**Review by aryx — APPROVED** [review]

File: src/optimizing/Analyze_pattern.ml (hunk 3 above)
Comment (IagoAbal) [reply src/optimizing/Analyze_pattern.ml]: TBH I don't think I benchmarked it, but I'll do now.

**Comment by IagoAbal** [comment]
Preliminary benchmarks look OK, just ~4% slow in average, a few repos were ~20% slower though. Not many differences in findings, etc. But I think we can fix that by in other ways, just prefer to leave that for subsequent PRs. I'll re-run the benchmarks now after having applied the review comments. If they're good I'll post them and merge.
